### ISSUE TITLE
ParallelNode not producing expected output

### ISSUE THREAD

**Issue by kchristidis** [onsi/ginkgo#439]
## Context

- `ginkgo` rev: `9008c7`
- `gomega` rev: `de89e6`

Consider this simple (and silly) test suite:

```go
func TestIntegration(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Integration Suite")
}

var _ = Describe("Parallelism", func() {
	foo := 0

	Specify(fmt.Sprintf("Node %d of %d should find the original value unchanged", config.GinkgoConfig.ParallelNode, config.GinkgoConfig.ParallelTotal), func() {
		Expect(foo).To(Equal(0))
	})

	Specify(fmt.Sprintf("Node %d of %d should find the original value unchanged", config.GinkgoConfig.ParallelNode, config.GinkgoConfig.ParallelTotal), func() {
		Expect(foo).To(Equal(0))
	})
})
```

I invoke it by  forcing the suite to run in 2 nodes, i.e.: `ginkgo -nodes=2 -v`

## Expectation

I expect to see references to "node 1" and "node 2" in the test output (when running in verbose mode).

## Reality

Instead both nodes show up with index 0:

```bash
$ ginkgo -nodes=2 -v
Running Suite: Integration Suite
================================
Random Seed: 1520643995
Will run 2 specs

Running in parallel across 2 nodes

Parallelism
  Node 0 of 0 should find the original value unchanged
  /Users/kchrist/Go/src/github.com/kchristidis/fo/integration/integration_suite_test.go:20
•
------------------------------
Parallelism
  Node 0 of 0 should find the original value unchanged
  /Users/kchrist/Go/src/github.com/kchristidis/foo/integration/integration_suite_test.go:24
•
Ran 2 of 2 Specs in 0.002 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 0 Skipped

Ginkgo ran 1 suite in 762.35331ms
Test Suite Passed
```

## Question

What am I missing?

**Comment by williammartin** [comment]
@kchristidis Sorry it took so much time to look at this. This was a tricksy one, but here's the summary:

^ I figured it out and it was related to the initialisation phases.

TL;DR:

When compiling a test (e.g. with `go test -c`, as ginkgo does), a `_testmain.go` file gets created which will create a new `M` struct (massive look of disapproval to Go here) via `NewMainStart()` and call `Run()` on it. The `Run()` method _sneakily_ parses flags (https://github.com/golang/go/blob/master/src/testing/testing.go#L976) that we set up in ginkgo's `init()` (https://github.com/onsi/ginkgo/blob/master/ginkgo_dsl.go#L54).

Now, looking at Golang's ordering for initialisation of your testing package at runtime we roughly get:

```
1. Start initialising the testing package (your code)
2. Initialise imports, specifically ginkgo (your import of ginkgo)
3. Run the `init()` func in the ginkgo package (https://github.com/onsi/ginkgo/blob/master/ginkgo_dsl.go#L53-L58), which sets up config flags.
4. Evaluate `var _ = Describe()` expression as part of your test package initialisation (this will evaluate the `fmt.Sprintf`)
5. Store spec blocks for later evaluation (https://github.com/onsi/ginkgo/blob/master/ginkgo_dsl.go#L396)
6. Run _testmain.go and parse flags (https://github.com/golang/go/blob/master/src/testing/testing.go#L976)
7. Run the stored ginkgo specs (your specify func)
```

We can see that in step 4 when config.GinkgoConfig.Parallel* are evaluated the flags haven't been parsed because the `main` phase hasn't run yet. By step 7, the flags have been parsed. Pretty unintuitive sorry :(

I don't think there's an easy way to resolve this. Is there a reason you really want this behaviour?

**Comment by kchristidis** [comment]
@williammartin: This is not a deal-breaker for me. Was mostly trying to figure out if I'm missing something. Thank you for explaining what's going on behind the scenes.

**Comment by williammartin** [comment]
OK cool, I'll try to get round to your other issue tomorrow. Thanks for the opportunity to understand this craziness!
